File: LeetCode/Java/0680. Valid Palindrome II.java
class Solution {
    public boolean validPalindrome(String s) {
        return skipFromStart(s) || skipFromEnd(s);
    }
    
    private boolean skipFromStart(String s) {
        int end = s.length() - 1;
        boolean skipped = false;
        
        for (int start = 0; start < end; start++) {
            if (s.charAt(start) != s.charAt(end) && !skipped) {
                skipped = true;
                end++;
            }
            else if (s.charAt(start) != s.charAt(end) && skipped)
                return false;
            
            end--;
        }
        return true;
    }
    
    private boolean skipFromEnd(String s) {
        int start = 0;
        boolean skipped = false;
        
        for (int end = s.length() - 1; end > start; end--) {
            if (s.charAt(start) != s.charAt(end) && !skipped) {
                skipped = true;
                start--;
            }
            else if (s.charAt(start) != s.charAt(end) && skipped)
                return false;
            
            start++;
        }
        return true;
    }
}

File: LeetCode/Python/0680. Valid Palindrome II.py
class Solution:
    def validPalindrome(self, s: str) -> bool:
        l, r = 0, len(s) - 1
        
        while l < r:
            if s[l] != s[r]:
                return self._validPalindrome(s, l + 1, r) or self._validPalindrome(s, l, r - 1)
            l += 1
            r -= 1
        
        return True
    
    def _validPalindrome(self, s, left, right):
        while left < right:
            if s[left] != s[right]:
                return False
            left += 1
            right -= 1
        return True
